keps/sig-windows/20190418-windows-security-context.md

```diff
@@ -0,0 +1,151 @@
+ 
+ 2. The Windows dockershim, CRI and OCI spec can already handle a username (instead of UID), which is interpreted inside the container to create a process as the intended user. This however is not surfaced as a field in the pod/container specs that an operator can specify. We want the ability to specify the desired username in the pod/container specs as fields and be able to pass them to the configured Windows runtime.
+
+### Goals
+
+Propose API enhancements in existing `PodSecurityContext` and `SecurityContext` structs for pods and individual containers respectively to allow operators to specify:
+- Name of a GMSA credential spec custom resource 
+- Full GMSA credential spec JSON 
+- A Windows username whose identity will be used to kick off the entrypoint process in containers.
+
+### Non-Goals
+
+- Details around GMSA end2end functionality and user stories as that is covered in the [GMSA](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md) KEP.
+- Implementation details and security considerations around how a GMSACredentialSpecName is expanded to GMSACredentialSpec as that is covered in details in the [GMSA](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md) KEP.
+- Details around how GMSA credential specs or Windows Username is passed through CRI and interpreted by Windows container run-times like Docker or ContainerD. Enhancements related to GMSA in CRI is already covered in the [GMSA](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md) KEP. Enhancements related to Username in CRI was introduced a while back in a [PR](https://github.com/kubernetes/kubernetes/pull/64009) 
+
+## Proposal
+
+In this KEP we propose a new field named `WindowsOptions` in the `PodSecurityContext` struct (associated with a pod spec) and the `SecurityContext` struct (associated with each container in a pod). `WindowsOptions` will be a pointer to a new struct of type `WindowsSecurityOptions`. The `WindowsSecurityOptions` struct will contain Windows OS specific security attributes scoped either at the pod level (applicable to all containers in the pod) or at the individual container level (that can override the pod level specification). Initially, the `WindowsSecurityOptions` struct will have the following fields:
+
+- GMSACredentialSpecName: A string specifying the name of a GMSA credential spec custom resource
+- GMSACredentialSpec: A string specifying the full credential spec JSON string associated with a GMSA credential spec
```

Comment:
How big is the json spec?  This might be difficult to format properly?

Reply:
Good point. In the GMSA KEP we detail how this field is auto populated by a webhook and not something the user needs to populate. Will clarify this.

Reply:
Is it needed in the public API then?  I am not too familiar, but I know there is an internal model that Kubernetes maintains and that might be the better location for it.  It seems confusing to have it on the public API surface if users do not use it.

Reply:
> Is it needed in the public API then? I am not too familiar, but I know there is an internal model that Kubernetes maintains and that might be the better location for it. It seems confusing to have it on the public API surface if users do not use it.

There's no private API used to communicate between components. The "internal" API types are used as a hub to convert between the various serialized forms, but all serialization (via the REST API or to/from etcd) uses public APIs.

Reply:
Additionally, the GMSA KEP (and current implementation) does call for users to be able to pass that info directly, even though the main use case should be to rely on the webhook for this.